circle-mlir/circle-mlir/lib/pass/src/ops/BatchNormalizationInferenceModeOp.h

```diff
@@ -0,0 +1,164 @@
+    assert(op.verify().succeeded());
+
+    mlir::Value input = adaptor.getX();
+    mlir::Value scale = adaptor.getScale();
+    mlir::Value bias = adaptor.getB();
+    mlir::Value mean = adaptor.getMean();
+    mlir::Value variance = adaptor.getVar();
+
+    mlir::Location opLoc = op->getLoc();
+
+    mlir::RankedTensorType intype = input.getType().dyn_cast_or_null<mlir::RankedTensorType>();
+    CHECK_VALID_RANK_2_4(intype);
+
+    mlir::RankedTensorType outtype = op.getType().dyn_cast_or_null<mlir::RankedTensorType>();
+    CHECK_VALID_RANK_2_4(outtype);
+
+    auto op_name = GetOperationName(op.getOperation());
+
+    float epsilon_val = op.getEpsilon().convertToFloat();
+    float momentum_val = op.getMomentum().convertToFloat();
+    LLVM_DEBUG({ llvm::dbgs() << "BatchNormIM name: " << op_name << "\n"; });
+    LLVM_DEBUG({ llvm::dbgs() << "BatchNormIM epsilon: " << epsilon_val << "\n"; });
+    LLVM_DEBUG({ llvm::dbgs() << "BatchNormIM momentum: " << momentum_val << "\n"; });
+
+    // NOTE below (a bit complicated) subgraph will be shrunk to Mul-Add by constant folding
+
+    // 0/ broadcast scale, bias, mean, variance to match 1xCx1x1
+    //    only rank 4 Constant works
+    if (intype.getRank() == 4)
+    {
+      scale = CreateConstBroadcastChn(rewriter, input, scale, op_name + "/scale");
+      bias = CreateConstBroadcastChn(rewriter, input, bias, op_name + "/bias");
+      mean = CreateConstBroadcastChn(rewriter, input, mean, op_name + "/mean");
+      variance = CreateConstBroadcastChn(rewriter, input, variance, op_name + "/variance");
+    }
+
+    // 1/ from onnx-tensorflow
+    //   running_mean = mean * momentum + mean * (1 - momentum)
+    //   running_variance = variance * momentum + variance * (1 - momentum)
+    //   inputs = [x, running_mean, running_variance, bias, scale]
+
+    mlir::Value const_momentum = CreateConst(rewriter, mean, momentum_val, op_name + "/momentum");
+    mlir::Value const_1_momentum =
+      CreateConst(rewriter, mean, 1.0f - momentum_val, op_name + "/1_momentum");
+    assert(not const_momentum.getType().isa<mlir::NoneType>());
+    assert(not const_1_momentum.getType().isa<mlir::NoneType>());
+
+    // mul_mm = mean * momentum
+    mlir::Location mul_mm_loc = mlir::NameLoc::get(rewriter.getStringAttr(op_name + "/mul_mm"));
+    mlir::Value mul_mm =
+      rewriter.create<MulOp>(mul_mm_loc, mean.getType(), mean, const_momentum, "NONE");
+    // mul_1_mm = mean * (1 - momentum)
+    mlir::Location mul_1_mm_loc = mlir::NameLoc::get(rewriter.getStringAttr(op_name + "/mul_1_mm"));
+    mlir::Value mul_1_mm =
+      rewriter.create<MulOp>(mul_1_mm_loc, mean.getType(), mean, const_1_momentum, "NONE");
+    // running_mean = mean * momentum + mean * (1 - momentum)
+    mlir::Location running_mean_loc =
+      mlir::NameLoc::get(rewriter.getStringAttr(op_name + "/running_mean"));
+    mlir::Value running_mean =
+      rewriter.create<AddOp>(running_mean_loc, mean.getType(), mul_mm, mul_1_mm, "NONE");
```

Comment:
It seems the `running_mean` is not used in the following codes.
We may comment out this lines.

```suggestion
    // only for completeness (not used)
    // // running_mean = mean * momentum + mean * (1 - momentum)
    // mlir::Location running_mean_loc =
    //   mlir::NameLoc::get(rewriter.getStringAttr(op_name + "/running_mean"));
    // mlir::Value running_mean =
    //   rewriter.create<AddOp>(running_mean_loc, mean.getType(), mul_mm, mul_1_mm, "NONE");
```

Reply:
@shs-park , wow... I haven't notice this -_-;
we may have to check why running_mean is not used from BN formula and original code.
until then, I'd like to post an issue for this land as-is.